README.md
# Faker Plugin

The **Faker** Plugin is an extension for [Grav CMS](http://github.com/getgrav/grav). Faker creates **dummy Grav page content** for testing and development purposes.

## Installation

Installing the Faker plugin can be done in one of three ways: The GPM (Grav Package Manager) installation method lets you quickly install the plugin with a simple terminal command, the manual method lets you do so via a zip file, and the admin method lets you do so via the Admin Plugin.

### GPM Installation (Preferred)

To install the plugin via the [GPM](http://learn.getgrav.org/advanced/grav-gpm), through your system's terminal (also called the command line), navigate to the root of your Grav-installation, and enter:

    bin/gpm install faker

This will install the Faker plugin into your `/user/plugins`-directory within Grav. Its files can be found under `/your/site/grav/user/plugins/faker`.

### Manual Installation

To install the plugin manually, download the zip-version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `faker`. You can find these files on [GitHub](https://github.com/trilbymedia/grav-plugin-faker) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/faker
	
> NOTE: This plugin is a modular component for Grav which may require other plugins to operate, please see its [blueprints.yaml-file on GitHub](https://github.com/trilbymedia/grav-plugin-faker/blob/master/blueprints.yaml).

### Instructions

This plugin provides a **CLI Command** to create dummy/fake data.  This simplest way to run **Faker** is to simply use:

```shell script
$ bin/plugin faker generate
```

You will be prompted for several options, all of which have defaults.  Just hit `[return]` to accept the defaults, or choose your own values:

```shell script
Generate Faker Content
======================

Number of nested_levels [1]:
Number of visible_levels [1]:
Number of items_per_level [100]:
Number of max_items [1000]:
Number of min_parts [5]:
Number of max_parts [20]:
Number of min_images [0]:
Number of max_images [3]:
Number of location [page://]:
```

Alternatively, you can set **any** or **all** the options via the CLI command itself:

```shell script
$ bin/plugin faker generate --nested-levels=3 --visible-levels=2 --items-per-level=10 --min-parts=5 --max-items=10000 --max-parts=20 --min-images=2 --max-images=5 --location=page://04.faker-pages
```

Get help with:

```shell script
$ bin/plugin faker help
```



CHANGELOG.md
# v0.1.0
##  08/20/2019

1. [](#new)
    * ChangeLog started...


faker.php
<?php
namespace Grav\Plugin;

use Grav\Common\Plugin;
use RocketTheme\Toolbox\Event\Event;

/**
 * Class FakerPlugin
 * @package Grav\Plugin
 */
class FakerPlugin extends Plugin
{
    /**
     * @return array
     *
     * The getSubscribedEvents() gives the core a list of events
     *     that the plugin wants to listen to. The key of each
     *     array section is the event that the plugin listens to
     *     and the value (in the form of an array) contains the
     *     callable (or function) as well as the priority. The
     *     higher the number the higher the priority.
     */
    public static function getSubscribedEvents()
    {
        return [
            ['autoload', 100000],
            'onPluginsInitialized' => ['onPluginsInitialized', 0]
        ];
    }

    /**
     * [onPluginsInitialized:100000] Composer autoload.
     *
     * @return ClassLoader
     */
    public function autoload()
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    /**
     * Initialize the plugin
     */
    public function onPluginsInitialized()
    {
        // Don't proceed if we are in the admin plugin
        if ($this->isAdmin()) {
            return;
        }
    }
}


classes/FakerMarkdownProvider.php
<?php

namespace Grav\Plugin\Faker;

use DavidBadura\MarkdownBuilder\MarkdownBuilder;
use Faker\Provider\Lorem;

/**
 * @author David Badura <d.a.badura@gmail.com>
 */
class FakerMarkdownProvider extends Lorem
{
    /**
     * @return string
     */
    public static function markdown($data, $images)
    {
        $parts = [];
        $items_count = self::numberBetween($data['min_parts'], $data['max_parts']);

        do {
            $parts[] = self::markdownH2();

            if (self::randomDigit() > 2) {
                $parts[] = self::markdownP(700);
            }

            if (self::randomDigit() > 4) {
                $parts[] = self::markdownBlockqoute();
            }

            if (self::randomDigit() > 4) {
                $parts[] = self::markdownBulletedList();
            }

            if (self::randomDigit() > 5) {
                $parts[] = self::markdownNumberedList();
            }

        } while (count($parts) < $items_count);

        foreach ($images as $image) {
            $parts[] = self::markdownInlineImg($image);
        }

        shuffle($parts);

        $parts = array_merge([self::markdownH1()], $parts);

        return implode("\n\n", $parts);
    }

    /**
     * @param int $maxNbChars
     * @return string
     */
    public static function markdownP($maxNbChars = 200)
    {
        return (new MarkdownBuilder())->p(self::text($maxNbChars))->getMarkdown();
    }

    /**
     * @param int $nbWords
     * @param bool $variableNbWords
     * @return string
     */
    public static function markdownH1($nbWords = 6, $variableNbWords = true)
    {
        return (new MarkdownBuilder())->h1(self::sentence($nbWords, $variableNbWords))->getMarkdown();
    }

    /**
     * @param int $nbWords
     * @param bool $variableNbWords
     * @return string
     */
    public static function markdownH2($nbWords = 6, $variableNbWords = true)
    {
        return (new MarkdownBuilder())->h2(self::sentence($nbWords, $variableNbWords))->getMarkdown();
    }

    /**
     * @param int $nbWords
     * @param bool $variableNbWords
     * @return string
     */
    public static function markdownH3($nbWords = 6, $variableNbWords = true)
    {
        return (new MarkdownBuilder())->h3(self::sentence($nbWords, $variableNbWords))->getMarkdown();
    }

    /**
     * @param int $maxNbChars
     * @return string
     */
    public static function markdownBlockqoute($maxNbChars = 200)
    {
        return (new MarkdownBuilder())->blockqoute(self::text($maxNbChars))->getMarkdown();
    }

    /**
     * @param int $nb
     * @return string
     */
    public static function markdownBulletedList($nb = 3)
    {
        return (new MarkdownBuilder())->bulletedList(self::sentences($nb))->getMarkdown();
    }

    /**
     * @param int $nb
     * @return string
     */
    public static function markdownNumberedList($nb = 3)
    {
        return (new MarkdownBuilder())->numberedList(self::sentences($nb))->getMarkdown();
    }

    /**
     * @param int $maxNbChars
     * @return string
     */
    public static function markdownCode($maxNbChars = 200)
    {
        return (new MarkdownBuilder())->code(self::text($maxNbChars))->getMarkdown();
    }

    /**
     * @return string
     */
    public static function markdownInlineCode()
    {
        return (new MarkdownBuilder())->inlineCode(self::word());
    }

    /**
     * @return string
     */
    public static function markdownInlineItalic()
    {
        return (new MarkdownBuilder())->inlineItalic(self::word());
    }

    /**
     * @return string
     */
    public static function markdownInlineBold()
    {
        return (new MarkdownBuilder())->inlineBold(self::word());
    }

    /**
     * @return string
     */
    public static function markdownInlineLink()
    {
        return (new MarkdownBuilder())->inlineLink('http://google.com', self::word());
    }

    /**
     * @return string
     */
    public static function markdownInlineImg($image = null)
    {
        return (new MarkdownBuilder())->inlineImg(
            $image,
            self::word()
        );
    }
}


cli/GenerateFakeContentCommand.php
<?php

/**
 * @package    Grav\Plugin\Login
 *
 * @copyright  Copyright (C) 2014 - 2017 RocketTheme, LLC. All rights reserved.
 * @license    MIT License; see LICENSE file for details.
 */

namespace Grav\Plugin\Console;

use Faker\Factory;
use Grav\Common\Filesystem\Folder;
use Grav\Common\Grav;
use Grav\Common\Inflector;
use Grav\Console\ConsoleCommand;
use Grav\Plugin\Faker\FakerMarkdownProvider;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Style\SymfonyStyle;
use Symfony\Component\Console\Question\Question;

/**
 * Class CleanCommand
 *
 * @package Grav\Console\Cli
 */
class GenerateFakeContentCommand extends ConsoleCommand
{
    protected $options;
    protected $data;
    protected $helper;
    protected $faker;
    protected $total_pages;

    /**
     * Configure the command
     */
    protected function configure()
    {
        $this
            ->setName('generate')
            ->addOption(
                'nested-levels',
                null,
                InputOption::VALUE_OPTIONAL,
                'Number of nested levels - default 1'
            )
            ->addOption(
                'visible-levels',
                null,
                InputOption::VALUE_OPTIONAL,
                'Visible Levels - default is 1'
            )
            ->addOption(
                'items-per-level',
                null,
                InputOption::VALUE_OPTIONAL,
                'Number of items per level - default 100'
            )
            ->addOption(
                'max-items',
                null,
                InputOption::VALUE_OPTIONAL,
                'Maximum number of items - default 10000'
            )
            ->addOption(
                'min-parts',
                null,
                InputOption::VALUE_OPTIONAL,
                'Minimum number of parts - default 5 parts'
            )
            ->addOption(
                'max-parts',
                null,
                InputOption::VALUE_OPTIONAL,
                'Maximum number of parts - default 20 parts'
            )
            ->addOption(
                'min-images',
                null,
                InputOption::VALUE_OPTIONAL,
                'Minimum number of images - default 0'
            )
            ->addOption(
                'max-images',
                null,
                InputOption::VALUE_OPTIONAL,
                'Maximum number of images - default 3'
            )
            ->addOption(
                'location',
                null,
                InputOption::VALUE_OPTIONAL,
                'Location - default is `page://'
            )

            ->setDescription('Generates fake content')
            ->setHelp('The <info>lookup-user</info> finds a user based on some data query.')
        ;
    }

    /**
     * @return int|null|void
     */
    protected function serve()
    {
        include __DIR__ . '/../vendor/autoload.php';

        $io = new SymfonyStyle($this->input, $this->output);

        $grav = Grav::instance();
        $grav->setup();

        $io->title('Generate Faker Content');

        $this->data = [
            'nested_levels'     => 1,
            'visible_levels'    => 1,
            'items_per_level'   => 100,
            'max_items'         => 1000,
            'min_parts'         => 5,
            'max_parts'         => 20,
            'min_images'        => 0,
            'max_images'        => 3,
            'location'          => 'page://',
        ];

        $this->helper = $this->getHelper('question');

        // Ask questions if option not provided
        foreach (array_keys($this->data) as $key) {
            $options_key = Inflector::hyphenize($key);
            $this->options[$key] = $this->input->getOption($options_key);
            $this->askQuestion($key);
        }

        $location = $grav['locator']->findResource($this->data['location']);
        $this->total_pages = 0;


        if ($location) {
            $io->warning(sprintf('Creating content in %s', $location));

            $this->faker = Factory::create();
            $this->faker->addProvider(new FakerMarkdownProvider($this->faker));

            $this->createContent($location);

            $io->success($this->total_pages . ' pages created');

        } else {
            $io->error(sprintf('Sorry, location: %s does not exists', $location));
        }
    }

    protected function createContent($location, $level = 1) {


        for ($i = 0; $i < $this->data['items_per_level']; $i++) {

            if ($this->total_pages >= $this->data['max_items']) {
                break;
            }

            $item_number = $i;
            $title = Inflector::titleize($this->faker->words(4, true));
            $slug = Inflector::hyphenize($title);
            $folder_name = $slug;

            $template = 'default';

            // Handle visibility
            if ($this->data['visible_levels'] > 0 &&  $level <= $this->data['visible_levels']) {
                $folder_name = str_pad($i+1, 2, '0', STR_PAD_LEFT) . '.' . $folder_name;
            }

            $folder = $location . '/' . $folder_name;

            Folder::create($folder);

            $images = [];
            $image_count = $this->faker->numberBetween($this->data['min_images'], $this->data['max_images']);

            for ($j = 0; $j < $image_count; $j++) {
                $images[] = $this->createImage($folder);
            }

            $content = $this->faker->markdown($this->data, $images);
            $markdown = "---\ntitle: $title\n---\n\n" . $content;

            file_put_contents($folder . '/' . $template . '.md', $markdown);

            $this->output->writeln('<white>' . ($this->total_pages + 1) . '</white> ' . $folder . ' <green>✔</green>');
            $this->total_pages++;

            // check if children need to be created
            if ($this->data['nested_levels'] > 0 &&  $level < $this->data['nested_levels']) {
                $this->createContent($folder, $level + 1);
            }
        }
    }

    protected function createImage($path)
    {
        $filename = Inflector::hyphenize($this->faker->words(2, true)) . '.jpg';
        $path = $path . '/' . $filename;
        $source = __DIR__ . '/../media/sample.jpg';

        if (file_exists($source)) {
            copy($source, $path);
            return $filename;
        }

        return 'filenotfound.jpg';
    }



    protected function askQuestion($key)
    {
        $type = gettype($this->data[$key]);


        if (!$this->options[$key]) {
            // Get username and validate
            $question = new Question("Number of <yellow>$key</yellow> <green>[{$this->data[$key]}]</green>: ", $this->data[$key]);
            if (in_array($type, ['integer','double'])) {
                $method = 'validate' . ucFirst($type);
            } else {
                $method = 'normalize' . ucFirst($type);
            }

            $this->$method($question);
            $this->data[$key] = $this->helper->ask($this->input, $this->output, $question);
        } else {
            $this->data[$key] = $this->options[$key];
        }
    }

    /**
     * @param $question  Question
     */
    protected function validateInteger($question)
    {
        $question->setValidator(function ($answer) {
            if (!is_numeric($answer)) {
                throw new \RuntimeException(
                    'This value must be an integer'
                );
            }

            return intval($answer);
        });
    }

    /**
     * @param $question  Question
     */
    protected function normalizeString($question)
    {
        $question->setNormalizer(function ($value) {
            // $value can be null here
            return $value ? trim($value) : '';
        });
    }

    /**
     * @param $question  Question
     */
    protected function validateString($question)
    {
        $question->setValidator(function ($answer) {
            return $answer;
        });
    }
}
